### ISSUE TITLE
Test worker freezes in EndPointQueue.take()

### ISSUE THREAD

**Issue by ekoutanov** [gradle/gradle#10365]
`GradleWorkerMain` intermittently hangs while executing a JUnit 4 test. 

### Expected Behavior
Test runner should complete all tests and exit successfully (or in error, if the unit tests fail).

### Current Behavior
The test appears to end successfully, but the `Test worker` thread of the  `worker.org.gradle.process.internal.worker.GradleWorkerMain` process is stuck in a WAITING state at `org.gradle.internal.remote.internal.hub.queue.EndPointQueue.take:49`.

The problem is intermittent; observable if `./gradlew` is run in a loop. 

### Context
We have some concurrency tests which warrant repeat unit test execution to ascertain correctness. Problem has not been observed in Gradle 5.5.1 or earlier; this isn't to say that it wasn't there.

### Steps to Reproduce 
The problem doesn't appear to be related to a specific project and might occur once in a hundred or so runs on a heavily loaded machine. For this particular run, the specific project under test was [https://github.com/obsidiandynamics/fulcrum](https://github.com/obsidiandynamics/fulcrum).

Gradle is run with:
```
./gradlew test --no-daemon --info --stacktrace
```

gradle.properties:
```
org.gradle.daemon=true
org.gradle.parallel=true
org.gradle.fork=true
org.gradle.jvmargs=--add-opens=java.base/java.lang=ALL-UNNAMED
org.gradle.configureondemand=true
```

When the worker hangs, we attach `jstack` to it and determine the location of the parked threads. See output below. Of specific interest is the `Test worker` stack trace. It's apparent that the tests have finished executing. Is there a reason why it wouldn't make progress?

```
[vagrant@vm4 tmp]$ jps -l
29332 org.gradle.wrapper.GradleWrapperMain
29380 org.gradle.launcher.daemon.bootstrap.GradleDaemon
30379 jdk.jcmd/sun.tools.jps.Jps
29436 worker.org.gradle.process.internal.worker.GradleWorkerMain

[vagrant@vm4 tmp]$ /usr/lib/jvm/jdk-11.0.2/bin/jstack 29436
2019-08-26 06:54:53
Full thread dump OpenJDK 64-Bit Server VM (11.0.2+9 mixed mode):

Threads class SMR info:
_java_thread_list=0x00007fbb5c001de0, length=12, elements={
0x00007fbba4012800, 0x00007fbba413d000, 0x00007fbba4141000, 0x00007fbba4153800,
0x00007fbba4156000, 0x00007fbba4158000, 0x00007fbba415a000, 0x00007fbba41ae800,
0x00007fbba436f800, 0x00007fbba4615800, 0x00007fbba4639800, 0x00007fbb5c001000
}

"main" #1 prio=5 os_prio=0 cpu=1674.19ms elapsed=33701.28s tid=0x00007fbba4012800 nid=0x72fd waiting on condition  [0x00007fbbad94e000]
   java.lang.Thread.State: TIMED_WAITING (parking)
at jdk.internal.misc.Unsafe.park(java.base@11.0.2/Native Method)
- parking to wait for  <0x00000000e03abcf0> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
at java.util.concurrent.locks.LockSupport.parkNanos(java.base@11.0.2/LockSupport.java:234)
at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(java.base@11.0.2/AbstractQueuedSynchronizer.java:2123)
at java.util.concurrent.ThreadPoolExecutor.awaitTermination(java.base@11.0.2/ThreadPoolExecutor.java:1454)
at org.gradle.internal.concurrent.ManagedExecutorImpl.stop(ManagedExecutorImpl.java:87)
at org.gradle.internal.concurrent.DefaultExecutorFactory$TrackedManagedExecutor.stop(DefaultExecutorFactory.java:80)
at org.gradle.internal.concurrent.ManagedExecutorImpl.stop(ManagedExecutorImpl.java:77)
at org.gradle.internal.remote.internal.hub.MessageHub.stop(MessageHub.java:223)
at org.gradle.internal.concurrent.CompositeStoppable.stop(CompositeStoppable.java:106)
- locked <0x00000000ffe1c200> (a org.gradle.internal.concurrent.CompositeStoppable)
at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection.stop(MessageHubBackedObjectConnection.java:150)
at org.gradle.internal.concurrent.CompositeStoppable.stop(CompositeStoppable.java:106)
- locked <0x00000000ffcfaa20> (a org.gradle.internal.concurrent.CompositeStoppable)
at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:153)
at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:70)
at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:68)
at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:73)

"Reference Handler" #2 daemon prio=10 os_prio=0 cpu=3.31ms elapsed=33701.25s tid=0x00007fbba413d000 nid=0x730b waiting on condition  [0x00007fbba85ad000]
   java.lang.Thread.State: RUNNABLE
at java.lang.ref.Reference.waitForReferencePendingList(java.base@11.0.2/Native Method)
at java.lang.ref.Reference.processPendingReferences(java.base@11.0.2/Reference.java:241)
at java.lang.ref.Reference$ReferenceHandler.run(java.base@11.0.2/Reference.java:213)

"Finalizer" #3 daemon prio=8 os_prio=0 cpu=0.50ms elapsed=33701.24s tid=0x00007fbba4141000 nid=0x730c in Object.wait()  [0x00007fbba84ac000]
   java.lang.Thread.State: WAITING (on object monitor)
at java.lang.Object.wait(java.base@11.0.2/Native Method)
- waiting on <0x00000000e0176c60> (a java.lang.ref.ReferenceQueue$Lock)
at java.lang.ref.ReferenceQueue.remove(java.base@11.0.2/ReferenceQueue.java:155)
- waiting to re-lock in wait() <0x00000000e0176c60> (a java.lang.ref.ReferenceQueue$Lock)
at java.lang.ref.ReferenceQueue.remove(java.base@11.0.2/ReferenceQueue.java:176)
at java.lang.ref.Finalizer$FinalizerThread.run(java.base@11.0.2/Finalizer.java:170)

"Signal Dispatcher" #4 daemon prio=9 os_prio=0 cpu=8.69ms elapsed=33701.22s tid=0x00007fbba4153800 nid=0x730f runnable  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread0" #5 daemon prio=9 os_prio=0 cpu=10285.16ms elapsed=33701.22s tid=0x00007fbba4156000 nid=0x7310 waiting on condition  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE
   No compile task

"C1 CompilerThread0" #6 daemon prio=9 os_prio=0 cpu=3022.38ms elapsed=33701.21s tid=0x00007fbba4158000 nid=0x7314 waiting on condition  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE
   No compile task

"Sweeper thread" #7 daemon prio=9 os_prio=0 cpu=38.04ms elapsed=33701.21s tid=0x00007fbba415a000 nid=0x7315 runnable  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Common-Cleaner" #8 daemon prio=8 os_prio=0 cpu=49.42ms elapsed=33701.09s tid=0x00007fbba41ae800 nid=0x731c in Object.wait()  [0x00007fbb8c355000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
at java.lang.Object.wait(java.base@11.0.2/Native Method)
- waiting on <no object reference available>
at java.lang.ref.ReferenceQueue.remove(java.base@11.0.2/ReferenceQueue.java:155)
- waiting to re-lock in wait() <0x00000000e01c1008> (a java.lang.ref.ReferenceQueue$Lock)
at jdk.internal.ref.CleanerImpl.run(java.base@11.0.2/CleanerImpl.java:148)
at java.lang.Thread.run(java.base@11.0.2/Thread.java:834)
at jdk.internal.misc.InnocuousThread.run(java.base@11.0.2/InnocuousThread.java:134)

"Service Thread" #10 daemon prio=9 os_prio=0 cpu=0.11ms elapsed=33700.45s tid=0x00007fbba436f800 nid=0x7325 runnable  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Test worker" #12 prio=5 os_prio=0 cpu=388253.79ms elapsed=33697.43s tid=0x00007fbba4615800 nid=0x7332 waiting on condition  [0x00007fbb7d4de000]
   java.lang.Thread.State: WAITING (parking)
at jdk.internal.misc.Unsafe.park(java.base@11.0.2/Native Method)
- parking to wait for  <0x00000000e03f0798> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
at java.util.concurrent.locks.LockSupport.park(java.base@11.0.2/LockSupport.java:194)
at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(java.base@11.0.2/AbstractQueuedSynchronizer.java:2081)
at org.gradle.internal.remote.internal.hub.queue.EndPointQueue.take(EndPointQueue.java:49)
at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:401)
at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@11.0.2/ThreadPoolExecutor.java:1128)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@11.0.2/ThreadPoolExecutor.java:628)
at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
at java.lang.Thread.run(java.base@11.0.2/Thread.java:834)

"/0:0:0:0:0:0:0:1:48556 to /0:0:0:0:0:0:0:1:41457 workers Thread 3" #14 prio=5 os_prio=0 cpu=28.16ms elapsed=33697.19s tid=0x00007fbba4639800 nid=0x7334 runnable  [0x00007fbb7d2dc000]
   java.lang.Thread.State: RUNNABLE
at sun.nio.ch.EPoll.wait(java.base@11.0.2/Native Method)
at sun.nio.ch.EPollSelectorImpl.doSelect(java.base@11.0.2/EPollSelectorImpl.java:120)
at sun.nio.ch.SelectorImpl.lockAndDoSelect(java.base@11.0.2/SelectorImpl.java:124)
- locked <0x00000000e03efff0> (a sun.nio.ch.Util$2)
- locked <0x00000000e03edd60> (a sun.nio.ch.EPollSelectorImpl)
at sun.nio.ch.SelectorImpl.select(java.base@11.0.2/SelectorImpl.java:141)
at org.gradle.internal.remote.internal.inet.SocketConnection$SocketInputStream.read(SocketConnection.java:182)
at com.esotericsoftware.kryo.io.Input.fill(Input.java:146)
at com.esotericsoftware.kryo.io.Input.require(Input.java:178)
at com.esotericsoftware.kryo.io.Input.readByte(Input.java:295)
at org.gradle.internal.serialize.kryo.KryoBackedDecoder.readByte(KryoBackedDecoder.java:81)
at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:64)
at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:52)
at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:81)
at org.gradle.internal.remote.internal.hub.MessageHub$ConnectionReceive.run(MessageHub.java:268)
at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@11.0.2/ThreadPoolExecutor.java:1128)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@11.0.2/ThreadPoolExecutor.java:628)
at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
at java.lang.Thread.run(java.base@11.0.2/Thread.java:834)

"Attach Listener" #38 daemon prio=9 os_prio=0 cpu=2.50ms elapsed=0.14s tid=0x00007fbb5c001000 nid=0x76d4 waiting on condition  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"VM Thread" os_prio=0 cpu=4963.74ms elapsed=33701.28s tid=0x00007fbba4134800 nid=0x7309 runnable  

"GC Thread#0" os_prio=0 cpu=316124.24ms elapsed=33701.32s tid=0x00007fbba4039800 nid=0x72ff runnable  

"GC Thread#1" os_prio=0 cpu=317777.64ms elapsed=33700.59s tid=0x00007fbb74001000 nid=0x7324 runnable  

"G1 Main Marker" os_prio=0 cpu=117.89ms elapsed=33701.32s tid=0x00007fbba404a800 nid=0x7300 runnable  

"G1 Conc#0" os_prio=0 cpu=291806.36ms elapsed=33701.32s tid=0x00007fbba404c000 nid=0x7302 runnable  

"G1 Refine#0" os_prio=0 cpu=3.85ms elapsed=33701.32s tid=0x00007fbba406d800 nid=0x7303 runnable  

"G1 Young RemSet Sampling" os_prio=0 cpu=6142.61ms elapsed=33701.32s tid=0x00007fbba406f800 nid=0x7304 runnable  
"VM Periodic Task Thread" os_prio=0 cpu=27732.44ms elapsed=33700.46s tid=0x00007fbba4379000 nid=0x7326 waiting on condition  

JNI global refs: 15, weak refs: 0
```

### Your Environment
CentOS7 on x86. OpenJDK 11.0.2. Gradle 5.6.

**Comment by ekoutanov** [comment]
Update: can confirm that this is also happening on Gradle 5.5.1

**Comment by big-guy** [comment]
@ekoutanov The stack trace doesn't have much to go on. When it hangs like this, if you kill the `GradleWorkerMain` process, does the build complete with any other information?

**Comment by ekoutanov** [comment]
@big-guy The build completes with no further output.

**Comment by simtel12** [comment]
We have a similar issue - intermittently (1-in-35ish) one of the test workers on CI will hang.  Note that we have only observed this issue when running with the Amazon Corretto [8] JDK.  When using the Oracle Java8 JDK, we have not observed this.

```app@ip-172-19-111-211:~/workspace/testing/gleach/Test Pipeline$ /usr/lib/jvm/java-1.8.0-amazon-corretto/bin/jps -l
2243 /tmp/remoting.jar
21845 org.jetbrains.kotlin.daemon.KotlinCompileDaemon
6553 sun.tools.jps.Jps
21659 org.gradle.launcher.daemon.bootstrap.GradleDaemon
30171 worker.org.gradle.process.internal.worker.GradleWorkerMain
21614 org.gradle.wrapper.GradleWrapperMain```

If the GradleWorkerMain process is killed at this point, the build will continue (and fail due "unable to connect to child process").

[jstack_threaddump_30171_workermain.txt](https://github.com/gradle/gradle/files/3674281/jstack_threaddump_30171_workermain.txt)

**Comment by simtel12** [comment]
Update: it's now been observed happening multiple times with the Oracle Java8 JDK

**Comment by Jonatha1983** [comment]
Hi we also facing this exactly issue: 
```
"/127.0.0.1:58390 to /127.0.0.1:58387 workers Thread 3@1639" prio=5 tid=0x10 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	  at sun.nio.ch.KQueueArrayWrapper.kevent0(KQueueArrayWrapper.java:-1)
	  at sun.nio.ch.KQueueArrayWrapper.poll(KQueueArrayWrapper.java:198)
	  at sun.nio.ch.KQueueSelectorImpl.doSelect(KQueueSelectorImpl.java:117)
	  at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
	  - locked <0x1209> (a sun.nio.ch.KQueueSelectorImpl)
	  - locked <0x120a> (a java.util.Collections$UnmodifiableSet)
	  - locked <0x120b> (a sun.nio.ch.Util$3)
	  at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
	  at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:101)
	  at org.gradle.internal.remote.internal.inet.SocketConnection$SocketInputStream.read(SocketConnection.java:185)
	  at com.esotericsoftware.kryo.io.Input.fill(Input.java:146)
	  at com.esotericsoftware.kryo.io.Input.require(Input.java:178)
	  at com.esotericsoftware.kryo.io.Input.readByte(Input.java:295)
	  at org.gradle.internal.serialize.kryo.KryoBackedDecoder.readByte(KryoBackedDecoder.java:81)
	  at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:64)
	  at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:52)
	  at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:81)
	  at org.gradle.internal.remote.internal.hub.MessageHub$ConnectionReceive.run(MessageHub.java:268)
	  at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	  at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	  at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	  at java.lang.Thread.run(Thread.java:748)
```

Running with or without intellij:

Running with gradle 6.0.1
IntelliJ IDEA 2019.3 (Ultimate Edition)
Build #IU-193.5233.57, built on November 20, 2019

Runtime version: 11.0.4+10-b520.11 x86_64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
macOS 10.14.6
GC: ParNew, ConcurrentMarkSweep
Memory: 4029M
Cores: 8

**Comment by goldyliang** [comment]
I had the same issue. The test hanged at some point which did not progress at all, during a long unit test launching soapUI.

The initial gradle version we used is very old (2.8), and the issue happens after we upgrade to 6.4. Then I tried to downgrade the gradle version to see if there is any regression. When I downgraded it to 5.1, I started to see below out of memory exceptions thrown and the test cases proceeded instead of hanging (even all of them showing the same exceptions):

```
 ... ... .integration.soapui.SoapUIIntegrationTests > .. .... ] FAILED
     java.lang.OutOfMemoryError at SoapUIIntegrationTests.groovy:213
. .. ...
... ... .integration.soapui.SoapUIIntegrationTests > ... ... ] FAILED
   java.lang.OutOfMemoryError 
... ...
```
The solution was to increase the heap maximum size from the test by adding below in build.gradle (since new gradle versions fork new test worker process, the gradle.options would not work)

tasks.withType(Test) {
  ...
    minHeapSize = "128m"
    maxHeapSize = "1024m"
  ...
}

It seems that the gradle version above 5.1 has a bug that the out-of-memory exception from the test worker does not get shown back to the main gradle console so there was no clue what happened.

**Comment by technoir42** [comment]
Similar problem here. This process got stuck:
```
/usr/lib/jvm/java-8-oracle/bin/java -Djava.awt.headless=true -Djava.security.manager=worker.org.gradle.process.internal.worker.child.BootstrapSecurityManager -Dorg.gradle.native=false -XX:+UseG1GC -XX:SoftRefLRUPolicyMSPerMB=250 -javaagent:build/tmp/expandedArchives/org.jacoco.agent-0.8.5.jar_6a2df60c47de373ea127d14406367999/jacocoagent.jar=destfile=build/jacoco/testDebugUnitTest.exec,append=true,inclnolocationclasses=false,dumponexit=true,output=file,jmx=false -Xmx2560m -Dfile.encoding=UTF-8 -Duser.country -Duser.language=en -Duser.variant -ea -cp /home/ubuntu/.gradle/caches/6.4.1/workerMain/gradle-worker.jar worker.org.gradle.process.internal.worker.GradleWorkerMain 'Gradle Test Executor 22'
```

```
"main" #1 prio=5 os_prio=0 tid=0x00007f868400d800 nid=0x7072 waiting on condition [0x00007f868bcf6000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x0000000720107070> (a java.util.concurrent.CountDownLatch$Sync)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:997)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
	at java.util.concurrent.CountDownLatch.await(CountDownLatch.java:231)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:72)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:46)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:132)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:70)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:68)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:73)
```

**Comment by bmuskalla** [comment]
I've experienced a similar hang (IJ running tests while a rebase was active) and it got stuck downloading a new gradle distribution (newer on the target branch) as one of the test workers was still active but never finished. This was with `6.7-20200905230036+0000-bin`

[threaddump-1599476102020.log](https://github.com/gradle/gradle/files/5182718/threaddump-1599476102020.log)
